### ISSUE TITLE
[JENKINS-7046] Block Build when downstream project is building

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-23#8365]
<p>When a downstream build is in progress, Hudson should prevent the upstream builds from occurring until the downstream build is complete.</p>

<p>For example:</p>

<p>CompA->CompB->CompC->CompD</p>

<p>If component C is building Hudson should block builds on A and B, until C has finished.  </p>

<p>It should then rebuild A, B and C before allowing D to proceed.</p>

<p>As the current system stands, if a user presses "build now" on component A when component C is building, A builds immediately.  This can be worked around in part by using the locks & latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>).  Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour. </p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=aggamemnon">aggamemnon</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046" target="_blank">Block Build when downstream project is building</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-01-12T09:36:47+00:00
<li><b>votes</b>: 21
<li><b>watchers</b>: 20
<li><b>imported</b>: 2025-11-23
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>When a downstream build is in progress, Hudson should prevent the upstream builds from occurring until the downstream build is complete.</p>

<p>For example:</p>

<p>CompA-&gt;CompB-&gt;CompC-&gt;CompD</p>

<p>If component C is building Hudson should block builds on A and B, until C has finished.  </p>

<p>It should then rebuild A, B and C before allowing D to proceed.</p>

<p>As the current system stands, if a user presses "build now" on component A when component C is building, A builds immediately.  This can be worked around in part by using the locks &amp; latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>).  Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour. </p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Dependant matrix builds across Linux and Solaris</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=141922&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-141922">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have already set "Block build when upstream project is building" and it doesn't provide the behaviour I described in my first comment.  Either it isn't intended to or it doesn't work!</p>
</pre>
</details>

</ul>
</details>

<p>I have already set "Block build when upstream project is building" and it doesn't provide the behaviour I described in my first comment.  Either it isn't intended to or it doesn't work!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142110&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142110">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Seems like 2 different things are discussed here.. willemv's last comment refers to "block build when upstream project is building" not working transitively.. I have also observed this behavior.  willemv, maybe file a new issue for this?  <a href="https://issues.jenkins.io/browse/JENKINS-5125" title="Projects building at the same time despite &quot;Block build when upstream is building&quot; option" class="issue-link" data-issue-key="JENKINS-5125">JENKINS-5125</a> and <a href="https://issues.jenkins.io/browse/JENKINS-5150" title="Race condition for feature &quot;Block build when upstream project is building&quot;" class="issue-link" data-issue-key="JENKINS-5150">JENKINS-5150</a> don't seem to specifically mention this..</p>

<p>However, it appears to me like the original request describes blocking in the opposite direction.. "block build when downstream project is building".  Hudson currently has no such feature.  Can you describe why you want this?  It doesn't make sense to me..  If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build.  So why block A at this point?  Might as well run it in parallel with C to get all the builds completed more quickly.</p>
</pre>
</details>

</ul>
</details>

<p>Seems like 2 different things are discussed here.. willemv's last comment refers to "block build when upstream project is building" not working transitively.. I have also observed this behavior.  willemv, maybe file a new issue for this?  <a href="https://issues.jenkins.io/browse/JENKINS-5125" title="Projects building at the same time despite "Block build when upstream is building" option" class="issue-link" data-issue-key="JENKINS-5125">JENKINS-5125</a> and <a href="https://issues.jenkins.io/browse/JENKINS-5150" title="Race condition for feature "Block build when upstream project is building"" class="issue-link" data-issue-key="JENKINS-5150">JENKINS-5150</a> don't seem to specifically mention this..</p>

<p>However, it appears to me like the original request describes blocking in the opposite direction.. "block build when downstream project is building".  Hudson currently has no such feature.  Can you describe why you want this?  It doesn't make sense to me..  If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build.  So why block A at this point?  Might as well run it in parallel with C to get all the builds completed more quickly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142138&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142138">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>"If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build. So why block A at this point? Might as well run it in parallel with C to get all the builds completed more quickly."</p>

<p>Comp C may depend on objects contained within A, hence if A rebuilds C will fail, or be inconsistent.</p>

<p>"Aggamemnon, which of the two do you actually need ?"</p>

<p>Both</p>
</pre>
</details>

</ul>
</details>

<p>"If we start the upstream job A, we know that will eventually trigger the downstream job C.. but since C has already started, there's nothing we can do at this point to avoid another build. So why block A at this point? Might as well run it in parallel with C to get all the builds completed more quickly."</p>

<p>Comp C may depend on objects contained within A, hence if A rebuilds C will fail, or be inconsistent.</p>

<p>"Aggamemnon, which of the two do you actually need ?"</p>

<p>Both</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>heymjo:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142277&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142277">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>mindless: imagine the case where you have a job that manipulates (e.g. destroys + rebuilds) a database schema. And then you have jobs that interact with this schema (unit,integration tests etc). Logically, the latter jobs have a dependency on the database job. We have expressed this dependency in the maven POMs, i.e. all application modules have a snapshot dependency on the database module.</p>

<p>ofcourse we don't want the database job to kick in during the building of any of its downstream jobs because it would just rip out the database schema while integration tests are in progress. That's why the database job would need a "block build when downstream project is building".</p>

<p>We can express this with locks and latches (urgh) or port allocator plugin but it would be a half baked job only and taking advantage of Hudson's snapshot dependency management (which is already <b>GREAT</b> by the way).</p>

<p>I hope this clarifies the need a bit for this feature.</p>

<p>Thanks
Jorg</p>
</pre>
</details>

</ul>
</details>

<p>mindless: imagine the case where you have a job that manipulates (e.g. destroys + rebuilds) a database schema. And then you have jobs that interact with this schema (unit,integration tests etc). Logically, the latter jobs have a dependency on the database job. We have expressed this dependency in the maven POMs, i.e. all application modules have a snapshot dependency on the database module.</p>

<p>ofcourse we don't want the database job to kick in during the building of any of its downstream jobs because it would just rip out the database schema while integration tests are in progress. That's why the database job would need a "block build when downstream project is building".</p>

<p>We can express this with locks and latches (urgh) or port allocator plugin but it would be a half baked job only and taking advantage of Hudson's snapshot dependency management (which is already <b>GREAT</b> by the way).</p>

<p>I hope this clarifies the need a bit for this feature.</p>

<p>Thanks<br/>
Jorg</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>heymjo:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=142278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-142278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>(i took the liberty to update the subject to make it less confusing)</p>
</pre>
</details>

</ul>
</details>

<p>(i took the liberty to update the subject to make it less confusing)</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dmichonneau:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143197&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143197">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here is a simple use case:</p>

<p>A: compilation of a jar
B: test the jar</p>

<p>B is running and locks the jar file
A is triggered: it fails to write the jar file after compilation.</p>
</pre>
</details>

</ul>
</details>

<p>Here is a simple use case:</p>

<p>A: compilation of a jar<br/>
B: test the jar</p>

<p>B is running and locks the jar file<br/>
A is triggered: it fails to write the jar file after compilation.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143210&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143210">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For above use case I recommend using copyartifact plugin to copy the jar file to the workspace of job B, and use that when running tests.  Job A can now run while B is already running for a previous build of the jar, and A/B are also more flexible in that they can run on different slaves.</p>
</pre>
</details>

</ul>
</details>

<p>For above use case I recommend using copyartifact plugin to copy the jar file to the workspace of job B, and use that when running tests.  Job A can now run while B is already running for a previous build of the jar, and A/B are also more flexible in that they can run on different slaves.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dmichonneau:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143218&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143218">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>That only moves the problem.</p>

<p>The jar copy is either done by A or B.</p>

<p>If done by A while B is running: the copy fails as B locks the file
If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p>

<p>Besides that, it's an intrusive approach on tests that are not designed to run on something else that the compilation output directory.</p>
</pre>
</details>

</ul>
</details>

<p>That only moves the problem.</p>

<p>The jar copy is either done by A or B.</p>

<p>If done by A while B is running: the copy fails as B locks the file<br/>
If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p>

<p>Besides that, it's an intrusive approach on tests that are not designed to run on something else that the compilation output directory.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143620&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143620">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>If done by A while B is running: the copy fails as B locks the file</p></blockquote>

<p>I don't see how a running B can lock an artifact from a previous build.  Hudson makes a copy of the file when it archives the artifact.. it wouldn't be copying a file that is in-use/locked by a newly running build of B, right?</p>

<blockquote><p>If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p></blockquote>

<p>B would never invoke a copy when A is configured with a copyartifact build step.</p>

<p>Here is more detail of what I was describing above:</p>
<ol>
	<li>A runs; Hudson archives an artifact</li>
	<li>B runs; it has a copyartifact build step to copy the artifact from the latest completed build of A.. if A is currently running that is ok; it is copying from the archived artifacts, not the workspace that is in use by that running build where files may be locked.</li>
	<li>A runs; if B is currently running that is ok since it is using a copy of a previous artifact.. this build can run to completion and archive a new artifact, then trigger a new build of B.</li>
</ol>


<p>If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal.</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>If done by A while B is running: the copy fails as B locks the file</p></blockquote>

<p>I don't see how a running B can lock an artifact from a previous build.  Hudson makes a copy of the file when it archives the artifact.. it wouldn't be copying a file that is in-use/locked by a newly running build of B, right?</p>

<blockquote><p>If done by B while A is running: the copy fails as A locks the file. There is a shorter interval during which the file is locked, but it can still happen.</p></blockquote>

<p>B would never invoke a copy when A is configured with a copyartifact build step.</p>

<p>Here is more detail of what I was describing above:</p>
<ol>
	<li>A runs; Hudson archives an artifact</li>
	<li>B runs; it has a copyartifact build step to copy the artifact from the latest completed build of A.. if A is currently running that is ok; it is copying from the archived artifacts, not the workspace that is in use by that running build where files may be locked.</li>
	<li>A runs; if B is currently running that is ok since it is using a copy of a previous artifact.. this build can run to completion and archive a new artifact, then trigger a new build of B.</li>
</ol>


<p>If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143666&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143666">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>"If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal."</p>

<p>We have a compilation output directory.  The issue is becoming more pressing.</p>
</pre>
</details>

</ul>
</details>

<p>"If you have a "compilation output directory" instead of a simple artifact, I can see how this is less than ideal."</p>

<p>We have a compilation output directory.  The issue is becoming more pressing.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=143921&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-143921">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?</p>
</pre>
</details>

</ul>
</details>

<p>Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144013&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144013">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>"Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?"</p>

<p>Not feasible due to the large size of the build directories, the large number of them and the extra unnecessary complexity this would add to the build environment.</p>
</pre>
</details>

</ul>
</details>

<p>"Maybe it's worth prototyping a setup where you zip up the output directory and archive that zip as an artifact, in case this issue is not resolved for some time.. though maybe it's very large and that's why you haven't tried that already?"</p>

<p>Not feasible due to the large size of the build directories, the large number of them and the extra unnecessary complexity this would add to the build environment.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144034&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144034">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Have you tried a setup with locks&amp;latches plugin?</p>
</pre>
</details>

</ul>
</details>

<p>Have you tried a setup with locks&latches plugin?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aggamemnon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144037&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144037">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Have you read the description? <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>"This can be worked around in part by using the locks &amp; latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>). Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour."</p>
</pre>
</details>

</ul>
</details>

<p>Have you read the description? <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>"This can be worked around in part by using the locks & latches plugin, but this has the side affect of disabling parallel Matrix builds which massively increases the build time (see <a href="https://issues.jenkins.io/browse/JENKINS-2725" title="locks and latches shouldnt' be global in multi-config" class="issue-link" data-issue-key="JENKINS-2725"><del>JENKINS-2725</del></a>). Further, it should NOT be necessary to use ANY plugins to provide this fundamental behaviour."</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>robsimon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144042&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144042">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>On top the "lock" is released before the artifacts are copied back and the next job in line is kicked off. That means this approach does not work.</p>
</pre>
</details>

</ul>
</details>

<p>On top the "lock" is released before the artifacts are copied back and the next job in line is kicked off. That means this approach does not work.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144044&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144044">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>yes, read it long ago..</p>
</pre>
</details>

</ul>
</details>

<p>yes, read it long ago..</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mleinart:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-7046?focusedId=144353&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-144353">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Here's another use case I have:</p>

<p>Large build with a large suite of tests (unit + 1.5hrs worth of selenium browser tests).</p>

<p>Job A: 
  Builds the app, starts it up
Job B:
  Matrix job using a custom workspace that points to Job A's workspace
  4 elements in 1 Axis: 4 test suites that are run simultaneously</p>

<p>We cant have Job A run while Job B is running because it'll wipe the workspace out.  Job A has an SCM poll trigger which on a busy day will already have another run queued while B is still running.  Rather than space SCM polls several hours apart, we'd rather the build for Job A be blocked by its downstream job.</p>

<p>Locks and latches doesn't work because <b>each</b> instance of Job B (it's a matrix job, remember) tries to grab a lock, preventing the instances from executing concurrently.</p>


<p>(Also note that we've found custom workspace to be the best option because the Copy Workspace has some crappy limitations (doesnt copy permissions, doesnt copy empty directories) and tarring/untarring a 1gb build is a waste of time and space)</p>
</pre>
</details>

</ul>
</details>

<p>Here's another use case I have:</p>

<p>Large build with a large suite of tests (unit + 1.5hrs worth of selenium browser tests).</p>

<p>Job A: <br/>
  Builds the app, starts it up<br/>
Job B:<br/>
  Matrix job using a custom workspace that points to Job A's workspace<br/>
  4 elements in 1 Axis: 4 test suites that are run simultaneously</p>

<p>We cant have Job A run while Job B is running because it'll wipe the workspace out.  Job A has an SCM poll trigger which on a busy day will already have another run queued while B is still running.  Rather than space SCM polls several hours apart, we'd rather the build for Job A be blocked by its downstream job.</p>

<p>Locks and latches doesn't work because <b>each</b> instance of Job B (it's a matrix job, remember) tries to grab a lock, preventing the instances from executing concurrently.</p>


<p>(Also note that we've found custom workspace to be the best option because the Copy Workspace has some crappy limitations (doesnt copy permissions, doesnt copy empty directories) and tarring/untarring a 1gb build is a waste of time and space)</p>
